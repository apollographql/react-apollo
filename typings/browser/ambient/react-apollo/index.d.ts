// Generated by typings
// Source: ambient.d.ts
declare module "lodash.isobject" {
  import main = require('~lodash/index');
  // export = main.isObject;
  public isObject = main.isObject;
}

declare module "lodash.isequal" {
  import main = require('~lodash/index');
  // export = main.isEqual;
  public isEqual = main.isEqual;
}