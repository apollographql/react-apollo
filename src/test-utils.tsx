import * as React from 'react';
import ApolloClient from 'apollo-client';
import { DefaultOptions, Resolvers } from 'apollo-client';
import { InMemoryCache as Cache } from 'apollo-cache-inmemory';

import { ApolloProvider } from './index';
import { MockedResponse, MockLink } from './test-links';
import { ApolloCache } from 'apollo-cache';
export * from './test-links';

export interface MockedProviderProps<TSerializedCache = {}> {
  mocks?: ReadonlyArray<MockedResponse>;
  addTypename?: boolean;
  defaultOptions?: DefaultOptions;
  cache?: ApolloCache<TSerializedCache>;
  resolvers?: Resolvers;
  childProps?: object;
}

export interface MockedProviderState {
  client: ApolloClient<any>;
}

export class MockedProvider extends React.Component<MockedProviderProps, MockedProviderState> {
  public static defaultProps: MockedProviderProps = {
    addTypename: true,
  };

  constructor(props: MockedProviderProps) {
    super(props);

    const {
      mocks,
      addTypename,
      defaultOptions,
      cache,
      resolvers,
    } = this.props;
    const client = new ApolloClient({
      cache: cache || new Cache({ addTypename }),
      defaultOptions,
      link: new MockLink(mocks || [], addTypename),
      resolvers,
    });

    this.state = { client };
  }

  public render() {
    const { childProps } = this.props;

    return (
      <ApolloProvider client={this.state.client}>
        {React.cloneElement(React.Children.only(this.props.children), { ...childProps })}
      </ApolloProvider>
    );
  }

  public componentWillUnmount() {
    // Since this.state.client was created in the constructor, it's this
    // MockedProvider's responsibility to terminate it.
    this.state.client.stop();
  }
}
