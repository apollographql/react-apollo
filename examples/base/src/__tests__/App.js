import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from 'react-apollo/test-utils';

import { HERO_QUERY, withCharacter, CharacterWithoutData, App } from '../App';

const query = HERO_QUERY;

import {
  empty,
  hero_no_friends,
  empty_array_friends,
  friend_without_appearsIn,
  full,
} from '../__mocks__/data';

class ErrorBoundary extends React.Component {
  componentDidCatch(e) {
    console.log(e);
  }

  render() {
    return this.props.children;
  }
}

const variables = { episode: 'NEWHOPE' };

describe('withCharacter', () => {
  it('shapes the props into variables', done => {
    class Container extends React.Component {
      componentWillMount() {
        try {
          expect(this.props.variables).toEqual(variables);
          done();
        } catch (e) {
          done.fail(e);
        }
      }
      render() {
        return null;
      }
    }

    const ContainerWithData = withCharacter(Container);
    const mocks = [{ request: { query, variables }, result: { data: { hero: empty } } }];
    renderer.create(
      <ErrorBoundary>
        <MockedProvider mocks={mocks}>
          <ContainerWithData {...variables} />
        </MockedProvider>
      </ErrorBoundary>,
    );
  });
  it('reshapes the data into the passed props', done => {
    class Container extends React.Component {
      componentWillReceiveProps(next) {
        try {
          expect(next.hero).toEqual(hero_no_friends);
          done();
        } catch (e) {
          done.fail(e);
        }
      }
      render() {
        return null;
      }
    }

    const ContainerWithData = withCharacter(Container);
    const mocks = [
      {
        request: { query, variables },
        result: { data: { hero: hero_no_friends } },
      },
    ];
    renderer.create(
      <ErrorBoundary>
        <MockedProvider mocks={mocks}>
          <ContainerWithData {...variables} />
        </MockedProvider>
      </ErrorBoundary>,
    );
  });
  it('has a loading state', done => {
    class Container extends React.Component {
      componentWillMount() {
        expect(this.props.loading).toBe(true);
      }
      componentWillReceiveProps(next) {
        expect(next.loading).toBe(false);
        done();
      }
      render() {
        return null;
      }
    }

    const ContainerWithData = withCharacter(Container);
    const mocks = [
      {
        request: { query, variables },
        result: { data: { hero: hero_no_friends } },
      },
    ];
    renderer.create(
      <ErrorBoundary>
        <MockedProvider mocks={mocks}>
          <ContainerWithData {...variables} />
        </MockedProvider>
      </ErrorBoundary>,
    );
  });
  it('has a potential error state', done => {
    class Container extends React.Component {
      componentWillMount() {
        expect(this.props.loading).toBe(true);
      }
      componentWillReceiveProps(next) {
        expect(next.loading).toBe(false);
        expect(next.error.message).toMatch(/these are not the droids you are looking for/);
        done();
      }
      render() {
        return null;
      }
    }

    const ContainerWithData = withCharacter(Container);
    const mocks = [
      {
        request: { query, variables },
        error: new Error('these are not the droids you are looking for'),
      },
    ];

    renderer.create(
      <ErrorBoundary>
        <MockedProvider mocks={mocks}>
          <ContainerWithData {...variables} />
        </MockedProvider>
      </ErrorBoundary>,
    );
  });
});

describe('CharacterWithoutData', () => {
  it('handles a loading state', () => {
    const output = renderer.create(<CharacterWithoutData loading />);
    expect(output.toJSON()).toMatchSnapshot();
  });
  it('handles an error state', () => {
    const output = renderer.create(<CharacterWithoutData error />);
    expect(output.toJSON()).toMatchSnapshot();
  });
  it('returns markup for null response', () => {
    const output = renderer.create(<CharacterWithoutData hero={empty} />);
    expect(output.toJSON()).toMatchSnapshot();
  });
  it('returns markup for a hero with no friends', () => {
    const output = renderer.create(<CharacterWithoutData hero={hero_no_friends} />);
    expect(output.toJSON()).toMatchSnapshot();
  });
  it('returns markup for empty array of friends', () => {
    const output = renderer.create(<CharacterWithoutData hero={empty_array_friends} />);
    expect(output.toJSON()).toMatchSnapshot();
  });
  it('returns markup for a friend without an appearsIn', () => {
    const output = renderer.create(<CharacterWithoutData hero={friend_without_appearsIn} />);
    expect(output.toJSON()).toMatchSnapshot();
  });
  it('renders a full data result', () => {
    const output = renderer.create(<CharacterWithoutData hero={full} />);
    expect(output.toJSON()).toMatchSnapshot();
  });
});

describe('App', () => {
  it('renders the data from NEWHOPE', () => {
    const mocks = [{ request: { query, variables }, result: { data: { hero: empty } } }];
    const output = renderer.create(
      <ErrorBoundary>
        <MockedProvider mocks={mocks}>
          <App />
        </MockedProvider>
      </ErrorBoundary>,
    );

    expect(output.toJSON()).toMatchSnapshot();
  });
});
