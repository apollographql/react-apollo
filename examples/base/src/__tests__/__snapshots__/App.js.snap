// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the data from NEWHOPE 1`] = `
<div>
  <div>
    Loading
  </div>
</div>
`;

exports[`CharacterWithoutData handles a loading state 1`] = `
<div>
  Loading
</div>
`;

exports[`CharacterWithoutData handles an error state 1`] = `
<h1>
  ERROR
</h1>
`;

exports[`CharacterWithoutData renders a full data result 1`] = `
<div>
  <div>
    <h3>
      r2d2
    </h3>
    <h6>
      luke
      :
       
      newhope
    </h6>
  </div>
</div>
`;

exports[`CharacterWithoutData returns markup for a friend without an appearsIn 1`] = `
<div>
  <div>
    <h3>
      r2d2
    </h3>
    <h6>
      luke
      :
       
      newhope
    </h6>
    <h6>
      james
      :
       
      
    </h6>
  </div>
</div>
`;

exports[`CharacterWithoutData returns markup for a hero with no friends 1`] = `
<div>
  <div>
    <h3>
      r2d2
    </h3>
  </div>
</div>
`;

exports[`CharacterWithoutData returns markup for empty array of friends 1`] = `
<div>
  <div>
    <h3>
      r2d2
    </h3>
  </div>
</div>
`;

exports[`CharacterWithoutData returns markup for null response 1`] = `<div />`;
